# <center> CSC17104-FINAL PROJECT </center>

## Mục lục
- [I. Thông tin nhóm](#i)
- [II. Thu thập dữ liệu](#ii)
    - [1. Chủ đề](#ii1)
    - [2. Nguồn dữ liệu](#ii2)
    - [3. Cách thu thập](#ii3)
- [III. Khám phá dữ liệu](#iii)
    - [1. Mô tả dữ liệu](#iii1)
    - [2. Tiền xử lí dữ liệu](#iii2)
- [IV. Đặt câu hỏi có ý nghĩa và trả lời](#iv)
    - [1. Câu hỏi 1](#iv1)
    - [2. Câu hỏi 2](#iv2)
    - [3. Câu hỏi 3](#iv3)
- [V. Nhận xét quá trình làm đồ án](#v)
    - [1. Thành viên 1](#v1)
    - [2. Thành viên 2](#v2)
    - [3. Thành viên 3](#v3)
    - [4. Ý kiến nhóm](#v4)
- [Tài liệu tham khảo](#tltk)

## I. Thông tin nhóm <a id="i"></a>

| Họ tên                | MSSV              | Công việc | Hoàn thành(%)|
|-----------------------|-------------------|-----------|--------------|
|Nguyễn Ngọc Minh Anh      |   22127023        |           | 100%         |
|Nguyễn Đức Hưng      |   22127142        |           | 100%         |
|Lê Huỳnh Nhật Hưng      |   22127140        |           | 100%         |

## II. Thu thập dữ liệu <a id="ii"></a>

### 1. Chủ đề <a id="ii1"></a>

Chúng em lựa chọn dataset về phim gồm các thông tin như đánh giá phim(ratings.csv), thể loại phim(movies.csv), nhãn(tags.csv).

### 2. Nguồn dữ liệu <a id="ii2"></a>

Dữ liệu mà chúng em sử dụng được lấy tập dữ liệu MovieLens từ trang GroupLens [link](https://grouplens.org/datasets/movielens/?fbclid=IwY2xjawHHjY5leHRuA2FlbQIxMAABHSNdVj6x86bIVx6UppaMq3AIINmbUwp3eTaUITt07DCsZfCtGnjp19zvNQ_aem_bWp2xc35td5tb09lVpc4yw). GroupLens nổi tiếng với các nghiên cứu về hệ thống gợi ý và cung cấp các tập dữ liệu phổ biến cho cộng đồng nghiên cứu.

Dữ liệu chúng em sử dụng được tác giả thu thập bằng cách chọn ngẫu nhiên người dùng để đưa vào. Tất cả người dùng được chọn đều đã đánh giá ít nhất 20 bộ phim trên website [link](https://movielens.org/). Không có thông tin nhân khẩu học nào được đưa vào. Mỗi người dùng được biểu thị bằng một id và không cung cấp thông tin nào khác. 

Dữ liệu này là dữ liệu được cập nhật theo thời gian để phù hợp với việc học, nghiên cứu nên có thể sẽ không còn đúng trong tương lai. Ở đây nhóm chúng em đã tải bản được cập nhật cuối cùng vào tháng 9 năm 2018.

Khi tải dataset này về đồng nghĩa với việc chúng em đã chấp nhận điều khoản sử dụng của GroupLens(không cho phép chia sẻ lại công khai, không dùng cho mục đích thương mại,...) đã được nêu trong phần đầu của tệp tin README.txt.

Ở đây, chúng em dùng dataset được khuyến khích cho học tập, chúng em lựa chọn tập có quy mô nhỏ với khoảng hơn 3MB. Và vì chúng em tập trung vào đặt câu hỏi, phân tích mối liên hệ giữa lượt đánh giá, thể loại phim, nhãn của phim nên có lẽ chúng em sẽ không xem xét đến tệp tin links.csv

### 3. Cách thu thập <a id="ii3"></a>

Chúng em tải tệp dữ liệu đã nén từ website của GroupLens mà tụi em đã đề cấp ở phần trên.

## III. Khám phá dữ liệu  <a id="iii"></a>

### 1. Mô tả dữ liệu <a id="iii1"></a>

Phần mô tả dữ liệu nằm trong Data/README.txt, và chúng em có tóm tắt lại như sau:
 
- ratings.csv:
    - userId: Id ngẫu nhiên được tạo để ẩn đi danh tính thật của người dùng đưa vào, định danh(trong dataset) cho mỗi người dùng.
    - movieId: Id để định danh phim.
    - rating: đánh giá trên thang 5, nhỏ nhất là 0.5 (0.5 - 5).
    - timestamp: số giây(khi thực hiện đánh giá) tính từ nửa đêm của Coordinated Universal Time (UTC) của  ngày 1 tháng 1 năm 1970.
- tags.csv:
    - userId: Id ngẫu nhiên được tạo để ẩn đi danh tính thật của người dùng đưa vào, định danh(trong dataset) cho mỗi người dùng.
    - movieId: Id để định danh phim.
    - tag: nhãn dán mà người dùng tìm trước khi chọn phim(1 dạng đánh giá của người dùng về phim).
    - timestamp: số giây(khi thực hiện đánh giá) tính từ nửa đêm của Coordinated Universal Time (UTC) của  ngày 1 tháng 1 năm 1970.
- movies.csv:
    - movieId: Id để định danh phim.
    - title: Tên phim
    - genre: Thể loại của phim

### 2. Tiền xử lí dữ liệu <a id="iii2"></a>

#### Đọc dữ liệu

Như đã đề cập, ở phạm vi đồ án này, chúng em sẽ không quan tâm đến dữ liệu trong tệp links.csv nên đầu tiên, chúng em sẽ đọc dữ liệu 3 tệp tin còn lại

In [1]:
import pandas as pd


In [2]:
movies = pd.read_csv('../Data/movies.csv')
movies.head(5)

movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy

In [3]:
ratings = pd.read_csv('../Data/ratings.csv')
ratings.head(5)

userId  movieId  rating  timestamp
0       1        1     4.0  964982703
1       1        3     4.0  964981247
2       1        6     4.0  964982224
3       1       47     5.0  964983815
4       1       50     5.0  964982931

In [4]:
tags = pd.read_csv('../Data/tags.csv')
tags.head(5)

userId  movieId              tag   timestamp
0       2    60756            funny  1445714994
1       2    60756  Highly quotable  1445714996
2       2    60756     will ferrell  1445714992
3       2    89774     Boxing story  1445715207
4       2    89774              MMA  1445715200

#### Hiểu dữ liệu

Sau khi đã đọc xong, chúng em sẽ bắt đầu xem xét dữ liệu về số dòng, cột, ý nghĩa của chúng,...

##### Dữ liệu có bao nhiêu dòng, bao nhiêu cột? Ý nghĩa của mỗi dòng, mỗi cột?

In [5]:
movies_shape = movies.shape
ratings_shape = ratings.shape
tags_shape = tags.shape
print("Movie:", movies_shape)
print("Rating:", ratings_shape)
print("Tag:", tags_shape)

Movie: (9742, 3)
Rating: (100836, 4)
Tag: (3683, 4)


Kích thước của các dataframe cũng khá vừa phải, cụ thể với:

- `ratings` (phần chính mà chúng ta sẽ quan tâm) có 100836 dòng và 4 cột. Với mỗi dòng là 1 đánh giá của người dùng cho 1 phim.
- `movies` có 9742 dòng và 3 cột. Với mỗi dòng chứa thông tin về 1 phim.
- `tags` có 3683 dòng và 3 cột. Với mỗi dòng chứa thông tin về nhãn của 1 phim

Ý nghĩa mỗi cột của các dataframe thì chúng em đã nêu ở phần mô tả dữ liệu.

##### Dữ liệu có trùng lặp không?

In [6]:
print("Số dòng trùng lặp của rating:", ratings.duplicated().sum())
print("Số dòng trùng lặp của movie:", movies.duplicated().sum())
print("Số dòng trùng lặp của tag:", tags.duplicated().sum())

Số dòng trùng lặp của rating: 0
Số dòng trùng lặp của movie: 0
Số dòng trùng lặp của tag: 0


Ở tất cả dataframe đều không có sự trùng lặp dữ liệu.

##### Dữ liệu có trống không?

In [7]:
print("Số dòng trống của rating:", ratings.isna().all(axis=1).sum())
print("Số dòng trống của movie:", movies.isna().all(axis=1).sum())
print("Số dòng trống của tag:", tags.isna().all(axis=1).sum())

Số dòng trống của rating: 0
Số dòng trống của movie: 0
Số dòng trống của tag: 0


Ở tất cả dataframe đều không có dòng trống

#### Tiếp theo, ta sẽ tìm hiểu, phân tích các khía cạnh sâu hơn của dữ liệu

##### Bắt đầu với việc tìm hiểu loại dữ liệu ở các cột

In [8]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100836 entries, 0 to 100835
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   userId     100836 non-null  int64  
 1   movieId    100836 non-null  int64  
 2   rating     100836 non-null  float64
 3   timestamp  100836 non-null  int64  
dtypes: float64(1), int64(3)
memory usage: 3.1 MB


In [9]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9742 entries, 0 to 9741
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   movieId  9742 non-null   int64 
 1   title    9742 non-null   object
 2   genres   9742 non-null   object
dtypes: int64(1), object(2)
memory usage: 228.5+ KB


In [10]:
tags.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3683 entries, 0 to 3682
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   userId     3683 non-null   int64 
 1   movieId    3683 non-null   int64 
 2   tag        3683 non-null   object
 3   timestamp  3683 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 115.2+ KB


Chúng em sẽ không tính toán với dữ liệu về Id nên em sẽ không quan tâm đến nó.

Dữ liệu timestamp thì em sẽ đổi về định dạng datetime để có thể khai thác các thông tin tiềm ẩn(như khoảng thời gian nào thì các phim thể loại nào được đánh giá nhiều,...).

Về thể loại phim, vì phụ vụ việc khai thác dữ liệu sau này, em sẽ đi sâu vào tìm hiểu cũng như chuyển đổi nó về dạng ma trận one hot(True False).

##### Chuyển đổi dữ liệu timestamp thành dạng datetime

In [11]:
ratings['Date'] = pd.to_datetime(ratings['timestamp'], unit='s', utc=True)
tags['Date'] = pd.to_datetime(tags['timestamp'], unit='s', utc=True)

ratings.drop(columns=['timestamp'], inplace = True)
tags.drop(columns=['timestamp'], inplace = True)

ratings.head(5)

userId  movieId  rating                      Date
0       1        1     4.0 2000-07-30 18:45:03+00:00
1       1        3     4.0 2000-07-30 18:20:47+00:00
2       1        6     4.0 2000-07-30 18:37:04+00:00
3       1       47     5.0 2000-07-30 19:03:35+00:00
4       1       50     5.0 2000-07-30 18:48:51+00:00

Sau khi chuyển đổi, em sẽ xóa cột timestamp ban đầu đi vì đã không cần đến nó nữa.

##### Khai thác sâu về cột genres

In [12]:
genres = set(movies['genres'].str.split('|').explode())
print(len(genres))
genres

20


{'(no genres listed)',
 'Action',
 'Adventure',
 'Animation',
 'Children',
 'Comedy',
 'Crime',
 'Documentary',
 'Drama',
 'Fantasy',
 'Film-Noir',
 'Horror',
 'IMAX',
 'Musical',
 'Mystery',
 'Romance',
 'Sci-Fi',
 'Thriller',
 'War',
 'Western'}

Ta nhận thấy có tất cả 19 thể loại(vì có giá trị sai là `(no genres listed)`). Tiếp theo ta sẽ xem thử có bao nhiêu phim không xác định được thể loại.

In [13]:
dropId = []
for i in range(len(movies)):
    if movies['genres'][i] == '(no genres listed)':
        dropId.append(movies['movieId'][i])
print(len(dropId))

34


Chỉ có 34 phim không xác định thể loại(1 con số khá khiêm tốn). Vì vậy, để đơn giản thì chúng em quyết định sẽ xóa các phim chưa xác định thể loại khỏi các dataframe.

In [14]:
movies = movies[~movies['movieId'].isin(dropId)]
tags = tags[~tags['movieId'].isin(dropId)]
ratings = ratings[~ratings['movieId'].isin(dropId)]

movies.shape

(9708, 3)

Ban đầu có 9742 phim, bây giờ còn 9708 phim, vậy là ta đã xóa 34 phim cần xóa.

Tiếp theo, ta sẽ chuyển các giá trị kiểu object của cột genres thành ma trận one hot để phục vụ việc tính toán, khai thác thông tin sau này.

In [15]:
movies['Genres_Split'] = movies['genres'].str.split('|')
genres_exploded = movies.explode('Genres_Split')

one_hot = pd.get_dummies(genres_exploded['Genres_Split'])
movies = movies.join(one_hot.groupby(one_hot.index).sum())
movies.drop(columns=['Genres_Split', 'genres'], inplace=True)

movies.head(5)

movieId                               title  Action  Adventure  Animation  \
0        1                    Toy Story (1995)       0          1          1   
1        2                      Jumanji (1995)       0          1          0   
2        3             Grumpier Old Men (1995)       0          0          0   
3        4            Waiting to Exhale (1995)       0          0          0   
4        5  Father of the Bride Part II (1995)       0          0          0   

   Children  Comedy  Crime  Documentary  Drama  ...  Film-Noir  Horror  IMAX  \
0         1       1      0            0      0  ...          0       0     0   
1         1       0      0            0      0  ...          0       0     0   
2         0       1      0            0      0  ...          0       0     0   
3         0       1      0            0      1  ...          0       0     0   
4         0       1      0            0      0  ...          0       0     0   

   Musical  Mystery  Romance  Sci-Fi  Thriller  War  Western  
0        0        0        0       0         0    0        0  
1        0        0        0       0         0    0        0  
2        0        0        1       0         0    0        0  
3        0        0        1       0         0    0        0  
4        0        0        0       0         0    0        0  

[5 rows x 21 columns]

Vậy ta đã thêm các cột ứng với ma trận one hot, 1 nếu đó là thể loại của phim, 0 nếu ngược lại. Cũng như xóa cột genres ban đầu.

##### Tiếp theo ta sẽ tìm hiểu các giá trị thống kê đối với dữ liệu

In [28]:
print(ratings.describe())
print('\n\n')
print(ratings['Date'].min(), "   ", ratings['Date'].max())

              userId        movieId         rating
count  100789.000000  100789.000000  100789.000000
mean      326.123307   19373.299299       3.501563
std       182.603725   35420.417225       1.042448
min         1.000000       1.000000       0.500000
25%       177.000000    1199.000000       3.000000
50%       325.000000    2990.000000       3.500000
75%       477.000000    8014.000000       4.000000
max       610.000000  193609.000000       5.000000



1996-03-29 18:36:55+00:00     2018-09-24 14:27:30+00:00


Với `ratings`, ta quan tâm đến vột rating, ta thấy các giá trị hoàn toàn bình thường: nhỏ nhất là 0.5, lớn nhất là 5(hoàn toàn phù hợp với mô tả dữ liệu), userId và movieId thì chỉ là các Id định danh nên ta sẽ không để tâm nhiều đến chúng, về dữ liệu ngày tháng, các giá trị min(max) của nó đều hoàn toàn bình thường

In [20]:
print(len(set(movies['title'])))
movies['title'].sample(5)

9703


3382            Gross Anatomy (a.k.a. A Cut Above) (1989)
2611                              Madame Sousatzka (1988)
6604                                        Saw IV (2007)
8593    Asterix: The Land of the Gods (Astérix: Le dom...
9491                                         CHiPS (2017)
Name: title, dtype: object

Mỗi phim có 1 tên khác nhau, điều này là hoàn toàn bình thường. Đối với ma trận one hot, vì chúng em mới xử lí, xây dựng nó và đã kiểm tra(có niềm tin rằng) nó sẽ không có sai.

In [30]:
print(len(set(tags['tag'])))
print('\n\n')
print(tags['tag'].sample(5))
print('\n\n')
print(tags['Date'].min(), '   ', tags['Date'].max())

1589



1820         Gambling
2819        Las Vegas
3200           clever
712         hilarious
3351    assassination
Name: tag, dtype: object



2006-01-13 19:09:12+00:00     2018-09-16 11:50:03+00:00


Các giá trị nhãn là do người dùng gán cho phim nên nó hoàn toàn có thể trùng lặp, 1589 giá trị riêng biệt là hoàn toàn bình thường. Các giá trị min(max) của ngày tháng cũng không có gì bất thường.

## IV. Đặt câu hỏi có ý nghĩa và trả lời  <a id="iv"></a>

### 1. Câu hỏi 1 <a id="iv1"></a>

### 2. Câu hỏi 2 <a id="iv2"></a>

### 3. Câu hỏi 3 <a id="iv3"></a>

## V. Nhận xét quá trình làm đồ án  <a id="v"></a>

### 1. Thành viên 1 <a id="v1"></a>

### 2. Thành viên 2 <a id="v2"></a>

### 3. Thành viên 3 <a id="v3"></a>

### 4. Ý kiến nhóm <a id="v4"></a>

## Tài liệu tham khảo  <a id="tltk"></a>